### PR TITLE
Add SIGINT handling

### PR DESCRIPTION
**Pull request by oetr**
This PR allows the fuzzer to gracefully shut down after the user has pressed CTRL-C. The coverage reports, sync/async suggestions, and registered custom hooks will be printed just like they should be after a regular shut down.

### CODE REVIEW

File: packages/core/core.ts [hunk 1]
```diff
@@ -165,18 +164,43 @@ function logInfoAboutFuzzerOptions(fuzzerOptions: string[]) {
 	});
 }
 
+function makeSigintCallback(options: Options, sync: boolean) {
```
Comment (bertschneider) [inline packages/core/core.ts:167]: As this function is only used in `startFuzzingNoInit` it could be changed to an inner function to be more local to its usage. I wonder if that would be cleaner.

File: packages/core/core.ts [hunk 2]
```diff
@@ -165,18 +164,43 @@ function logInfoAboutFuzzerOptions(fuzzerOptions: string[]) {
...
+	return () => {
+		stopFuzzing(
+			undefined,
+			options.expectedErrors,
+			options.coverageDirectory,
+			options.coverageReporters,
+			options.sync
+		);
+		// In case no error was thrown, we need to stop the native fuzzer manually.
+		stopNativeFuzzer(0);
```
Comment (bertschneider) [inline packages/core/core.ts:178]: `stopFuzzing` should always calls `Fuzzer.stopFuzzing[Async]` internally.

File: packages/fuzzer/utils.cpp [hunk 3]
```diff
@@ -19,6 +19,16 @@
 
 #define btoa(x) ((x) ? "true" : "false")
 
+bool sigintDetected = false;
+
+bool getSigintCalled() { return sigintDetected; }
+
+void sigintHandler(int signum) {
```
Comment (bertschneider) [inline packages/fuzzer/utils.cpp:26]: Could you add a comment describing why the signal handler callback saved in `fuzzing_sync` is not executed directly? As far as I understand the code, the flag would still be needed to block further fuzz target executions during shutdown.

File: packages/core/core.ts (hunk 2 above)
Comment (oetr) [reply packages/core/core.ts:178]: I wondered the same thing too, but the fuzzer does not exit without this! Looking at the code superficially, we only stop libfuzzer in case of an error on the JavaScript side, and simply return on no error?

File: packages/fuzzer/utils.cpp (hunk 3 above)
Comment (oetr) [reply packages/fuzzer/utils.cpp:26]: It was some weird node "V8" "wrong context"(? not sure about this one) ... something error when that function was called from the handler directly. So I moved the call to the function that already calls a node function without problems. Could it be that "env" variable that's missing?

File: tests/helpers.js [hunk 4]
```diff
@@ -27,6 +27,7 @@ const assert = require("assert");
 // `fuzzEntryPoint` (which would return a "1")
 const FuzzingExitCode = "77";
 const JestRegressionExitCode = "1";
+const WindowsExitCode = "1";
```
[the remark this answers is not in the document]
Comment (oetr) [reply tests/helpers.js:30]: So far I have been abusing `JestRegressionExitCode` for Windows exit codes on timeout and some other edge cases (exit codes on Windows and Linux/macOS differe sometimes). Here I wanted a dedicated name for exit code 1 just for windows, and in the next step I wanted to change all tests on windows that use that code.

File: tests/signal-handlers/signal-handlers.test.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+			.verbose(true)
+			.jestTestFile("tests.fuzz.js")
+			.jestTestName("^Jest Fuzzing and SIGINT$")
+			.jestRunInFuzzingMode(true)
+			.runs(1000)
+			.build();
+
+		fuzzTest.execute();
+
+		expect(fuzzTest.stdout).toContain("kill with SIGINT");
```
[the remark this answers is not in the document]
Comment (oetr) [reply tests/signal-handlers/signal-handlers.test.js:85]: I usually start extraction when a third duplicate arrives, which is just one test away here :smile: Probably a good idea to do so.

File: packages/fuzzer/utils.h [hunk 6]
```diff
@@ -15,10 +15,14 @@
 #pragma once
 
 #include <napi.h>
+#include <csignal>
 // Definitions from compiler-rt, including libfuzzer's entrypoint and the
 // sanitizer runtime initialization function.
 #include <fuzzer/FuzzerDefs.h>
 
+void sigintHandler(int sig);
```
[the remark this answers is not in the document]
Comment (oetr) [reply packages/fuzzer/utils.h:23]: Good idea! I had it there initially, but moved it to also cover the async case, which didn't work. Forgot to move the code back.

File: packages/fuzzer/utils.h [hunk 7]
```diff
@@ -15,10 +15,14 @@
 #pragma once
 
 #include <napi.h>
+#include <csignal>
```
[the remark this answers is not in the document]
Comment (oetr) [reply packages/fuzzer/utils.h:18]: I am not sure either, but an ok either way!

File: packages/core/core.ts (hunk 1 above)
Comment (oetr) [reply packages/core/core.ts:167]: With a comment explaing the locality, I agree!

[the remark this answers is not in the document]
Comment (oetr) [reply packages/core/core.ts:167]: Right, should change!

File: packages/fuzzer/utils.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (bertschneider) [reply packages/fuzzer/utils.h]: Moved to `fuzzing_sync.cpp`.

File: packages/core/core.ts (hunk 2 above)
Comment (bertschneider) [reply packages/core/core.ts]: Ah, yes, you're right. If no error was raised and none was expected the process can exit normally after libFuzzer finishes. I added an optional parameter to enforce a shutdown with a dedicated exit code.

File: packages/fuzzer/utils.cpp (hunk 3 above)
Comment (bertschneider) [reply packages/fuzzer/utils.cpp]: Added a comment.

File: tests/helpers.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (bertschneider) [reply tests/helpers.js:30]: Ah, I see. Then we can leave it there.

File: packages/fuzzer/utils.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (bertschneider) [reply packages/fuzzer/utils.h]: Moved.

**Review by br-lewis — APPROVED** [review]
